[AIRFLOW-510] Filter Paused Dags, Add Last Run Column and Trigger Dag Icon

btallman opened the pull request:
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- _https://issues.apache.org/jira/browse/AIRFLOW-510_

Per Apache guidelines you need to create a [Jira issue](https://issues.apache.org/jira/browse/AIRFLOW/).

Modify the HomeView to filter out paused dags if wanted and display the last
run date time on each dag, as well as allow externally triggering the dag from an icon

![homeview 1](https://cloud.githubusercontent.com/assets/173950/19336243/a9137ece-90be-11e6-9acb-ca74a93a28a7.png)
![homeview 2](https://cloud.githubusercontent.com/assets/173950/19336242/a90e4b0c-90be-11e6-963d-d75c872a93f9.png)

r39132 commented on airflow/models.py:
```diff
@@ -2671,17 +2671,22 @@ def normalize_schedule(self, dttm):
+    def get_last_dagrun(self, session=None, include_externally_triggered=False):
         """
         Returns the last dag run for this dag, None if there was none.
         Last dag run can be any type of run eg. scheduled or backfilled.
```
how about triggered or manual?

r39132 commented on airflow/models.py:
```diff
@@ -2671,17 +2671,22 @@ def normalize_schedule(self, dttm):
-        ).order_by(DR.execution_date.desc()).first()
+        )
+        if include_externally_triggered is False:
+            qry = qry.filter(DR.external_trigger == False)
```
Why exclude externally triggered dags here?

btallman replied:
That is the existing behavior. Just left the default the way it is.

r39132 commented on airflow/www/templates/airflow/dags.html:
```diff
@@ -102,66 +105,93 @@
+                        {% set last_run = dag.get_last_dagrun(include_externally_triggered=True) %}
+                        {% if last_run %}
+                            <a href="{{ url_for('airflow.graph', dag_id=last_run.dag_id, execution_date=last_run.execution_date ) }}">
+                                {{ last_run.start_date.strftime("%Y-%m-%d %H:%M") }}
```
Are these in GMT? If so, consider including the GMT timezone "Z" suffix.

r39132 commented on airflow/www/templates/airflow/dags.html:
```diff
@@ -102,66 +105,93 @@
                 <!-- Tree -->
                 <a href="{{ url_for('airflow.tree', dag_id=dag.dag_id, num_runs=25) }}" title="Tree View">
-                    <span class="glyphicon glyphicon-tree-deciduous" aria-hidden="true" title="Tree View"></span>
+                    <span class="glyphicon glyphicon-tree-deciduous" aria-hidden="true"></span>
```
Why have the titles been removed for all of these - this applies to all of the glyphicons below.

r39132 commented on airflow/www/views.py:
```diff
@@ -963,6 +963,42 @@ def run(self):
             "it should start any moment now.".format(ti))
         return redirect(origin)
 
+    @expose('/trigger')
```
This function is very similar to trigger_dag in the CLI. This is why we need an API, so common functionality can be used between the CLI and Web. Do you have an idea of how to reuse this code as per DRY? The main differences are flash vs logging calls.

btallman replied:
I think for now, better not to start building out an API, but continue to follow the existing pattern. If it helps, I used that code as a starting place.

Happy and excited to consider working on that API ASAP, but not for this version.

r39132 commented on airflow/configuration.py:
```diff
@@ -255,6 +255,10 @@ def run_command(command):
 # while fetching logs from other worker machine
 log_fetch_timeout_sec = 5
 
+# By default, the webserver shows paused DAGs. Flip this to hide paused
```
You also need to add this to TEST_CONFIG in the same file. That would allow someone (you?) to write tests for it in the future and ensure there is no regression, etc....

btallman committed:
Filter Paused Dags, Add Last Run Column and Trigger Dag Icon

Modify the HomeView to filter out paused dags if wanted (and a config value to set
the default), display the last run datetime on each dag, and allow externally triggering
the dag from an icon
